From: IrfanMohammad <irfan@graphiant.com>
Date: Thu, 27 Jul 2023 12:13:45 +0000
Subject: Cache sr_mod_oper_sub_t shm during pull callback

 To access the xpath of the oper sub, we held on to ext remap read lock
 which is not ideal, as it leads to a deadlock as described in ENG-3841.

 The relevant steps here for operational pull data are

 Step 1. oper sub lock read
 Step 2. ext remap lock read, read the evpipe, xpath of the oper sub
 Step 3. invoke callback
   Step 3a. sr_process_events (sr_module_oper_data_dup_enabled) ext WRITE lock
   ^^^ DEADLOCKED ^^^
   Step 3b. sr_process_events ext WRITE unlock
 Step 4. ext remap unlock read
 Step 5. oper sub unlock read

 Normally Step 3a requires only a read lock. However, if ext shm size
 has changed, due to another subscription add/del, we will need to remap it.
 This will require a write lock, and since read lock is already held in
 Step 2 this deadlocks.

 This can be prevented by caching `sr_mod_oper_sub_t` and unlocking
 ext_shm in Step 2. With this change the behavior is

 Step 1. oper sub lock read
 Step 2. ext remap lock read, cache mod_oper_sub info, unlock ext shm
 Step 3. invoke callback
   Step 3a. sr_process_events (sr_module_oper_data_dup_enabled) ext WRITE lock
   Step 3b. sr_process_events ext WRITE unlock
 Step 4. oper sub unlock read

Also fix wrong assumption in new_wrlock about cleared events
---
 src/modinfo.c   | 85 +++++++++++++++++++++++++++++++++++++++------------------
 src/shm_sub.c   | 19 +++++++------
 src/shm_sub.h   |  5 ++--
 src/shm_types.h |  9 ++++++
 src/subscr.c    |  6 ++--
 5 files changed, 83 insertions(+), 41 deletions(-)

diff --git a/src/modinfo.c b/src/modinfo.c
index 0938adc..6c76406 100644
--- a/src/modinfo.c
+++ b/src/modinfo.c
@@ -1065,8 +1065,8 @@ cleanup:
 static sr_error_info_t *
 sr_xpath_oper_data_get(struct sr_mod_info_mod_s *mod, const char *xpath, const char **request_xpaths,
         uint32_t req_xpath_count, const char *orig_name, const void *orig_data, uint32_t operation_id,
-        sr_mod_oper_get_sub_t *shm_subs, uint32_t idx1, const struct lyd_node *parent, uint32_t timeout_ms,
-        sr_conn_ctx_t *conn, struct lyd_node **oper_data)
+        sr_mod_oper_get_sub_cached_t *cached_subs, uint32_t idx1, const struct lyd_node *parent,
+        uint32_t timeout_ms, sr_conn_ctx_t *conn, struct lyd_node **oper_data)
 {
     sr_error_info_t *err_info = NULL, *cb_err_info = NULL;
     struct lyd_node *parent_dup = NULL, *last_parent;
@@ -1110,7 +1110,7 @@ sr_xpath_oper_data_get(struct sr_mod_info_mod_s *mod, const char *xpath, const c
 
     /* get data from client */
     if ((err_info = sr_shmsub_oper_get_notify(mod, xpath, request_xpath, parent_dup, orig_name, orig_data, operation_id,
-            shm_subs, idx1, timeout_ms, conn, oper_data, &cb_err_info))) {
+            cached_subs, idx1, timeout_ms, conn, oper_data, &cb_err_info))) {
         sr_errinfo_merge(&err_info, cb_err_info);
         goto cleanup;
     }
@@ -1214,6 +1214,17 @@ cleanup:
     return err_info;
 }
 
+static void
+sr_mod_free_cached_sub(sr_mod_oper_get_sub_cached_t *subs, size_t count)
+{
+    size_t i;
+    for (i = 0; i < count; i++) {
+        free(subs[i].xpath);
+        free(subs[i].xpath_subs);
+    }
+    free(subs);
+}
+
 /**
  * @brief Callback for merging operational data.
  */
@@ -1453,6 +1464,8 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
     int required, merged;
     struct ly_set *set = NULL;
     struct lyd_node *oper_data;
+    sr_mod_oper_get_sub_cached_t *cached_subs = NULL;
+    size_t oper_get_sub_count = 0, count = 0;
 
     if (!(get_oper_opts & SR_OPER_NO_STORED)) {
         /* process stored operational data */
@@ -1481,31 +1494,52 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
 
     /* EXT READ LOCK */
     if ((err_info = sr_shmext_conn_remap_lock(conn, SR_LOCK_READ, 0, __func__))) {
-        goto cleanup_opergetsub_unlock;
+        goto cleanup;
     }
+    oper_get_sub_count = mod->shm_mod->oper_get_sub_count;
+    cached_subs = calloc(oper_get_sub_count, sizeof *cached_subs);
 
     /* XPaths are ordered based on depth */
     shm_subs = (sr_mod_oper_get_sub_t *)(conn->ext_shm.addr + mod->shm_mod->oper_get_subs);
-    for (i = 0; i < mod->shm_mod->oper_get_sub_count; ++i) {
-        sub_xpath = conn->ext_shm.addr + shm_subs[i].xpath;
-        xpath_subs = (sr_mod_oper_get_xpath_sub_t *)(conn->ext_shm.addr + shm_subs[i].xpath_subs);
 
+    for (i = 0; i < oper_get_sub_count; ++i) {
         /* useless to retrieve configuration data, state data */
         if (((shm_subs[i].sub_type == SR_OPER_GET_SUB_CONFIG) && (get_oper_opts & SR_OPER_NO_CONFIG)) ||
                 ((shm_subs[i].sub_type == SR_OPER_GET_SUB_STATE) && (get_oper_opts & SR_OPER_NO_STATE))) {
             continue;
         }
+        count = shm_subs[i].xpath_sub_count;
+        assert(conn->ext_shm.addr + shm_subs[i].xpath);
+        cached_subs[i].xpath = strdup(conn->ext_shm.addr + shm_subs[i].xpath);
+        cached_subs[i].xpath_subs = malloc(count * sizeof *cached_subs[i].xpath_subs);
+        memcpy(cached_subs[i].xpath_subs, conn->ext_shm.addr + shm_subs[i].xpath_subs, count * sizeof *cached_subs[i].xpath_subs);
+        cached_subs[i].xpath_sub_count = count;
+    }
+
+    /* prevent use after ext remap unlock */
+    shm_subs = NULL;
+
+    /* EXT READ UNLOCK */
+    sr_shmext_conn_remap_unlock(conn, SR_LOCK_READ, 0, __func__);
+
+    for (i = 0; i < oper_get_sub_count; ++i) {
+        sub_xpath = cached_subs[i].xpath;
+        xpath_subs = cached_subs[i].xpath_subs;
+        if (!xpath_subs) {
+            /* Previously filtered */
+            continue;
+        }
 
         if (mod->xpath_count) {
             /* check whether these data are even required */
             for (j = 0; j < mod->xpath_count; ++j) {
                 if ((err_info = sr_xpath_oper_data_required(mod->xpaths[j], sub_xpath, &required))) {
-                    goto cleanup_opergetsub_ext_unlock;
+                    goto cleanup;
                 }
                 if (required) {
                     /* remember all xpaths causing these data to be required */
                     request_xpaths = sr_realloc(request_xpaths, (req_xpath_count + 1) * sizeof *request_xpaths);
-                    SR_CHECK_MEM_GOTO(!request_xpaths, err_info, cleanup_opergetsub_ext_unlock);
+                    SR_CHECK_MEM_GOTO(!request_xpaths, err_info, cleanup);
                     request_xpaths[req_xpath_count] = mod->xpaths[j];
                     ++req_xpath_count;
                 }
@@ -1519,13 +1553,13 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
 
         /* remove any present data */
         if (!(xpath_subs[0].opts & SR_SUBSCR_OPER_MERGE) && (err_info = sr_lyd_xpath_complement(data, sub_xpath))) {
-            goto cleanup_opergetsub_ext_unlock;
+            goto cleanup;
         }
 
         if (!(get_oper_opts & SR_OPER_NO_POLL_CACHED)) {
             /* try to get data from the cache */
             if ((err_info = sr_module_oper_data_update_cached(mod, sub_xpath, conn, data, &merged))) {
-                goto cleanup_opergetsub_ext_unlock;
+                goto cleanup;
             }
             if (merged) {
                 /* we have the data */
@@ -1535,7 +1569,7 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
 
         /* trim the last node to get the parent */
         if ((err_info = sr_xpath_trim_last_node(sub_xpath, &parent_xpath))) {
-            goto cleanup_opergetsub_ext_unlock;
+            goto cleanup;
         }
 
         if (parent_xpath) {
@@ -1545,7 +1579,7 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
             }
 
             if ((err_info = sr_lyd_find_xpath(*data, parent_xpath, &set))) {
-                goto cleanup_opergetsub_ext_unlock;
+                goto cleanup;
             }
 
             if (!set->count) {
@@ -1557,27 +1591,27 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
             if (xpath_subs[0].opts & SR_SUBSCR_OPER_ONCE) {
                 /* get oper data from the client */
                 if ((err_info = sr_xpath_oper_data_get(mod, sub_xpath, request_xpaths, req_xpath_count, orig_name,
-                        orig_data, operation_id, shm_subs, i, NULL, timeout_ms, conn, &oper_data))) {
-                    goto cleanup_opergetsub_ext_unlock;
+                        orig_data, operation_id, cached_subs, i, NULL, timeout_ms, conn, &oper_data))) {
+                    goto cleanup;
                 }
 
                 /* merge into one data tree */
                 if ((err_info = sr_lyd_merge(data, oper_data, 1, LYD_MERGE_DESTRUCT))) {
                     lyd_free_all(oper_data);
-                    goto cleanup_opergetsub_ext_unlock;
+                    goto cleanup;
                 }
             } else {
                 for (j = 0; j < set->count; ++j) {
                     /* get oper data from the client */
                     if ((err_info = sr_xpath_oper_data_get(mod, sub_xpath, request_xpaths, req_xpath_count, orig_name,
-                            orig_data, operation_id, shm_subs, i, set->dnodes[j], timeout_ms, conn, &oper_data))) {
-                        goto cleanup_opergetsub_ext_unlock;
+                            orig_data, operation_id, cached_subs, i, set->dnodes[j], timeout_ms, conn, &oper_data))) {
+                        goto cleanup;
                     }
 
                     /* merge into one data tree */
                     if ((err_info = sr_lyd_merge(data, oper_data, 1, LYD_MERGE_DESTRUCT))) {
                         lyd_free_all(oper_data);
-                        goto cleanup_opergetsub_ext_unlock;
+                        goto cleanup;
                     }
                 }
             }
@@ -1591,13 +1625,13 @@ next_iter:
         } else {
             /* top-level data */
             if ((err_info = sr_xpath_oper_data_get(mod, sub_xpath, request_xpaths, req_xpath_count, orig_name,
-                    orig_data, operation_id, shm_subs, i, NULL, timeout_ms, conn, &oper_data))) {
-                goto cleanup_opergetsub_ext_unlock;
+                    orig_data, operation_id, cached_subs, i, NULL, timeout_ms, conn, &oper_data))) {
+                goto cleanup;
             }
 
             if ((err_info = sr_lyd_merge(data, oper_data, 1, LYD_MERGE_DESTRUCT))) {
                 lyd_free_all(oper_data);
-                goto cleanup_opergetsub_ext_unlock;
+                goto cleanup;
             }
         }
 
@@ -1606,14 +1640,11 @@ next_iter:
         req_xpath_count = 0;
     }
 
-cleanup_opergetsub_ext_unlock:
-    /* EXT READ UNLOCK */
-    sr_shmext_conn_remap_unlock(conn, SR_LOCK_READ, 0, __func__);
-
-cleanup_opergetsub_unlock:
+cleanup:
     /* OPER GET SUB READ UNLOCK */
     sr_rwunlock(&mod->shm_mod->oper_get_lock, SR_SHMEXT_SUB_LOCK_TIMEOUT, SR_LOCK_READ, conn->cid, __func__);
 
+    sr_mod_free_cached_sub(cached_subs, oper_get_sub_count);
     free(request_xpaths);
     free(parent_xpath);
     ly_set_free(set, NULL);
diff --git a/src/shm_sub.c b/src/shm_sub.c
index 377b0f9..5e3360d 100644
--- a/src/shm_sub.c
+++ b/src/shm_sub.c
@@ -374,6 +374,7 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
         ret = sr_cond_clockwait(&sub_shm->lock.cond, &sub_shm->lock.mutex, COMPAT_CLOCK_ID, &timeout_abs);
     }
 
+    last_event = ATOMIC_LOAD_RELAXED(sub_shm->event);
     if (!sub_shm->lock.readers[0]) {
         /* FAKE WRITE LOCK */
         assert(!sub_shm->lock.writer);
@@ -381,15 +382,17 @@ sr_shmsub_notify_new_wrlock(sr_sub_shm_t *sub_shm, const char *shm_name, sr_sub_
 
         if (ret == ETIMEDOUT) {
             /* try to recover the event again in case the originator crashed later */
-            sr_shmsub_recover(sub_shm);
-            if (!ATOMIC_LOAD_RELAXED(sub_shm->event)) {
-                /* recovered */
+            if (last_event) {
+                sr_shmsub_recover(sub_shm);
+                last_event = ATOMIC_LOAD_RELAXED(sub_shm->event);
+            }
+            if (!last_event) {
+                /* recovered or the originator cleared the event just as we see a timeout */
                 ret = 0;
             }
         }
     }
 
-    last_event = ATOMIC_LOAD_RELAXED(sub_shm->event);
     last_request_id = ATOMIC_LOAD_RELAXED(sub_shm->request_id);
 
     if (ret) {
@@ -518,7 +521,6 @@ _sr_shmsub_notify_wait_wr(sr_sub_shm_t *sub_shm, sr_sub_event_t event, uint32_t
             if (event != last_event) {
                 SR_LOG_WRN("EV ORIGIN: SHM event \"%s\" ID %" PRIu32 " changed to \"%s\" unexpectedly",
                         sr_ev2str(event), operation_id, sr_ev2str(last_event));
-
             }
             SR_ERRINFO_COND(&err_info, __func__, ret);
         }
@@ -2193,7 +2195,7 @@ cleanup:
 sr_error_info_t *
 sr_shmsub_oper_get_notify(struct sr_mod_info_mod_s *mod, const char *xpath, const char *request_xpath,
         const struct lyd_node *parent, const char *orig_name, const void *orig_data, uint32_t operation_id,
-        sr_mod_oper_get_sub_t *oper_get_subs, uint32_t idx1, uint32_t timeout_ms, sr_conn_ctx_t *conn,
+        sr_mod_oper_get_sub_cached_t *cached_subs, uint32_t idx1, uint32_t timeout_ms, sr_conn_ctx_t *conn,
         struct lyd_node **data, sr_error_info_t **cb_err_info)
 {
     sr_error_info_t *err_info = NULL;
@@ -2203,14 +2205,15 @@ sr_shmsub_oper_get_notify(struct sr_mod_info_mod_s *mod, const char *xpath, cons
     char *parent_lyb = NULL;
     struct lyd_node *oper_data;
     sr_cid_t cid;
+    uint32_t cached_sub_count = cached_subs[idx1].xpath_sub_count;
 
     if (!request_xpath) {
         request_xpath = "";
     }
     cid = conn->cid;
 
-    for (i = 0; i < oper_get_subs[idx1].xpath_sub_count; i++) {
-        xpath_sub = &((sr_mod_oper_get_xpath_sub_t *)(conn->ext_shm.addr + oper_get_subs[idx1].xpath_subs))[i];
+    for (i=0; i < cached_sub_count; i++) {
+        xpath_sub = &(cached_subs[idx1].xpath_subs[i]);
 
         /* check subscription aliveness */
         if (!sr_conn_is_alive(xpath_sub->cid)) {
diff --git a/src/shm_sub.h b/src/shm_sub.h
index 6d818ba..1b82a5c 100644
--- a/src/shm_sub.h
+++ b/src/shm_sub.h
@@ -177,9 +177,8 @@ sr_error_info_t *sr_shmsub_change_notify_change_abort(struct sr_mod_info_s *mod_
  * @return err_info, NULL on success.
  */
 sr_error_info_t *sr_shmsub_oper_get_notify(struct sr_mod_info_mod_s *mod, const char *xpath, const char *request_xpath,
-        const struct lyd_node *parent, const char *orig_name, const void *orig_data, uint32_t operation_id,
-        sr_mod_oper_get_sub_t *oper_get_subs, uint32_t idx1, uint32_t timeout_ms, sr_conn_ctx_t *conn,
-        struct lyd_node **data, sr_error_info_t **cb_err_info);
+        const struct lyd_node *parent, const char *orig_name, const void *orig_data, uint32_t operation_id, sr_mod_oper_get_sub_cached_t *oper_get_subs,
+        uint32_t idx1, uint32_t timeout_ms, sr_conn_ctx_t *conn, struct lyd_node **data, sr_error_info_t **cb_err_info);
 
 /**
  * @brief Notify about (generate) an RPC/action event.
diff --git a/src/shm_types.h b/src/shm_types.h
index ecdcf8d..b466a04 100644
--- a/src/shm_types.h
+++ b/src/shm_types.h
@@ -256,6 +256,15 @@ typedef struct {
     uint32_t xpath_sub_count;   /**< Number of subscriptions for given XPath */
 } sr_mod_oper_get_sub_t;
 
+/**
+ * @brief cached oper get subscriptions so we don't need to hold remap lock
+ * */
+typedef struct {
+    sr_mod_oper_get_xpath_sub_t *xpath_subs; /**< XPath sub array */
+    uint32_t xpath_sub_count;   /**< Number of subscriptions for given XPath */
+    char *xpath;                /**< XPath of the subscription */
+} sr_mod_oper_get_sub_cached_t;
+
 /**
  * @brief Ext SHM module operational poll subscription.
  */
diff --git a/src/subscr.c b/src/subscr.c
index c618bb6..3699b42 100644
--- a/src/subscr.c
+++ b/src/subscr.c
@@ -46,7 +46,7 @@ sr_subscr_change_sub_add(sr_subscription_ctx_t *subscr, uint32_t sub_id, sr_sess
     uint32_t i;
     void *mem[4] = {NULL};
     int new_sub = 0;
-    sr_multi_sub_shm_t *multi_sub_shm;
+    sr_sub_shm_t *sub_shm;
 
     /* just to prevent problems in future changes */
     assert(has_subs_lock == SR_LOCK_WRITE);
@@ -111,8 +111,8 @@ sr_subscr_change_sub_add(sr_subscription_ctx_t *subscr, uint32_t sub_id, sr_sess
     /* new subscription */
     subscr->last_sub_id = sub_id;
 
-    multi_sub_shm = (sr_multi_sub_shm_t *)(change_sub->sub_shm.addr);
-    *is_event_pending = ATOMIC_LOAD_RELAXED(multi_sub_shm->subscriber_count) != 0;
+    sub_shm = (sr_sub_shm_t *)(change_sub->sub_shm.addr);
+    *is_event_pending = ATOMIC_LOAD_RELAXED(sub_shm->subscriber_count) != 0;
 
     return NULL;
 
